src/credentials/GroupDataProviderImpl.cpp

```diff
@@ -0,0 +1,1388 @@
+    VerifyOrReturnError(kInvalidEndpointId != fabric.first_endpoint, false);
+
+    // Existing fabric
+
+    EndpointData endpoint(fabric_index, fabric.first_endpoint);
+    size_t endpoint_count = 0;
+
+    // Loop through the fabric's endpoints
+    do
+    {
+        VerifyOrReturnError(CHIP_NO_ERROR == endpoint.Load(mStorage), false);
+        if (endpoint.id == mapping.endpoint)
+        {
+            // Target endpoint found
+            break;
+        }
+        endpoint.id = endpoint.next;
+    } while ((++endpoint_count < fabric.endpoint_count) && (kInvalidEndpointId != endpoint.id));
+    VerifyOrReturnError(endpoint.id == mapping.endpoint, false);
+
+    // Target endpoint found
+
+    GroupData group(fabric_index, mapping.endpoint, endpoint.first_group);
+
+    // Loop through the endpoint's groups
+    do
+    {
+        // Load next group
+        VerifyOrReturnError(CHIP_NO_ERROR == group.Load(mStorage), false);
+        if (group.group == mapping.group)
+        {
+            // Target group found
+            return true;
+        }
+        group.group = group.next;
+    } while (kUndefinedGroupId != group.group);
+
+    // Not found
+    return false;
+}
+
+CHIP_ERROR GroupDataProviderImpl::AddGroupMapping(chip::FabricIndex fabric_index, const GroupMapping & mapping)
+{
+    VerifyOrReturnError(mInitialized, CHIP_ERROR_INTERNAL);
+    VerifyOrReturnError(kInvalidEndpointId != mapping.endpoint, CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrReturnError(IsFabricGroupId(mapping.group), CHIP_ERROR_INVALID_ARGUMENT);
+
+    FabricData fabric(fabric_index);
+
+    if (CHIP_NO_ERROR != fabric.Load(mStorage) || kInvalidEndpointId == fabric.first_endpoint)
+    {
+        // New fabric, or new endpoint
+        GroupData(fabric_index, mapping.endpoint, mapping.group, mapping.name).Save(mStorage);
+        EndpointData(fabric_index, mapping.endpoint, mapping.group).Save(mStorage);
+        fabric.first_endpoint = mapping.endpoint;
+        fabric.endpoint_count = 1;
+        return fabric.Save(mStorage);
+    }
+
+    // Existing fabric
+
+    EndpointData endpoint(fabric_index, fabric.first_endpoint);
+    size_t endpoint_count = 0;
+
+    // Loop through the fabric's endpoints
+    do
+    {
+        if (CHIP_NO_ERROR != endpoint.Load(mStorage))
+        {
+            // Endpoint info not found
+            break;
+        }
+        if (endpoint.id == mapping.endpoint)
+        {
+            // Target endpoint info found
+            break;
+        }
+        endpoint.id = endpoint.next;
+    } while ((++endpoint_count < fabric.endpoint_count) && (kInvalidEndpointId != endpoint.id));
+
+    if (endpoint.id != mapping.endpoint)
+    {
+        // Endpoint not found, create new, and insert first
+        GroupData(fabric_index, mapping.endpoint, mapping.group, mapping.name).Save(mStorage);
+        EndpointData(fabric_index, mapping.endpoint, mapping.group, fabric.first_endpoint).Save(mStorage);
```

Comment:
Error handling must be done here and above.

That is, what happens if we run out of flash adding the above? The consistency of the data structures will be lost and we will have forever-orphan entries possibly

Reply:
That's why the order here is important. If the group cannot be stored, then the endpoint is not updated. If the endpoint storage fails, then the fabric is not updated, thus minimizing the risk.
An orphan entry is problematic, but may eventually be overwritten without problem. A `next` field pointing to an non-existing entry seems worse, and harder to recover from.